### ISSUE TITLE
(node:2135) UnhandledPromiseRejectionWarning: TypeError: Cannot cancel a stream that already has a reader

### ISSUE THREAD

**Issue by vballa** [openpgpjs/openpgpjs#783]
`openpgp.decrypt` resulting in the following unhandled error 

> (node:2135) UnhandledPromiseRejectionWarning: TypeError: Cannot cancel a stream that already has a reader
>     at e.r.cancel (/node_modules/openpgp/dist/openpgp.js:3:18264)
>     at Object.cancel (/node_modules/openpgp/dist/openpgp.js:22370:18)
>     at Message.decrypt (/node_modules/openpgp/dist/openpgp.js:32839:28)

Am I missing something in using this library? 

Here is my code

 

>  const options = {
>     message: await openpgp.message.readArmored(encrypted),    
>     publicKeys: (await openpgp.key.readArmored(pubkey)).keys, 
>     privateKeys: [privKeyObj],
>   }
>   const output = (await openpgp.decrypt(options)).data;

Please shed some light on this.

**Comment by twiss** [comment]
Could you turn on debugging by setting `openpgp.config.debug = true` before all your other code, and then post any other errors that show up in the console?

**Comment by vballa** [comment]
@twiss thanks for your reply. but i don't see any log statement even after setting 

`openpgp.config.debug = true;`

**Comment by vballa** [comment]
Output from --list-packets using gpg command line.

`gpg: encrypted with 4096-bit RSA key, ID <NUMBER>, created 2018-01-23`

is this what you meant? @twiss 

btw, i've replaced ID above, not sure if that is secure :)

**Comment by spearmootz** [comment]
I also get this error, funny thing is when i run the files individually they work but its crashing when i process multiple, even though i dont think im processing them at the same time (i wait for one to finish before i start the next)

**Comment by twiss** [comment]
Hmm, weird. Could you post your code?

**Comment by spearmootz** [comment]
this is my code basically, its typed.

first is the my library built on top of yours, and then the second is my process in which i use my library

```

import configuration from './configuration';
// @ts-ignore
import * as openpgp from 'openpgp';

export default async (data: Buffer) => {
  const privateKeys = await openpgp.key.readArmored(
    configuration.encryptionPrivateKey
  );

  const privateKey = privateKeys.keys[0];
  await privateKey.decrypt(configuration.encryptionPassword);

  return openpgp.decrypt({
    privateKeys: [privateKey],
    message: await openpgp.message.read(data),
    streaming: false
  });
};

```

```
const results = folderWatcher(incomingFolder, { currentFiles: true }).pipe(
      filter(
        fileName =>
          fileName.includes('.pgp') && fileName.includes('controlNumber')
      ),
      distinct(),
      concatMap(async (fileName: string) => {
        const filePath = path.join(incomingFolder, fileName);
        const buffer = await fs.readFile(filePath);
        const decryptedFile = (await pgpDecrypt(buffer)).data;
        const results = parser(decryptedFile);

        return results;
      })
    );
```

**Comment by spearmootz** [comment]
the funny thing is that i cannot reproduce it, it just sort of happens.

**Comment by twiss** [comment]
What happens if you comment out that line of code (dist/openpgp.js:32839)? Do you get a different error?

https://github.com/openpgpjs/openpgpjs/blob/c3419e5cd07e873a32f18bcd5a1636ad980e58ac/src/message.js#L132

**Comment by vballa** [comment]
I've commented dist/openpgp.js:32839 (L312 in the source) and error not being thrown anymore. and no new (different) errors either.

**Comment by twiss** [comment]
Hmm. That's good, but I'd prefer not to just remove that line without at least some idea of what's going on.

Could you perhaps try uncommenting it again and then changing the following version to just "0.1.0",

https://github.com/openpgpjs/openpgpjs/blob/c3419e5cd07e873a32f18bcd5a1636ad980e58ac/package.json#L74

and then running `npm install` and `grunt browserify:openpgp` and then testing again?

**Comment by spearmootz** [comment]
I read the code yesterday and went ahead and commented that line.but I was not able to trigger the bug, I passed more than 300 files with and without the line :(
Reproducing is the hard part

**Comment by vballa** [comment]
Tested with 

>  "@mattiasbuelens/web-streams-polyfill": "0.1.0",

Same error at Line number 32834

_webStreamTools2.default.cancel(symEncryptedPacket.encrypted); // Don't keep copy of encrypted data in memory.

**Comment by spearmootz** [comment]
i will comment out that line of code and have it run for a very long time, lets see what happens

**Comment by twiss** [comment]
@vballa Do you always get an error, or is it intermittent? Is it just for that message/key or others too? You might not be having the same issue as @spearmootz.

**Comment by vballa** [comment]
@twiss - i see the error consistently. tested with different message but still the same

**Comment by twiss** [comment]
@vballa Could you post a full example reproducing the error including an example key and message?

**Comment by twiss** [comment]
Also, do you happen to have `openpgp.config.integrity_protect = false`, `openpgp.config.ignore_mdc_error = true` or any other config options set?

**Comment by vballa** [comment]
@twiss i got this

`  openpgp.config.ignore_mdc_error = true;`

but not integrity_protect set

**Comment by twiss** [comment]
@vballa Ok, that does explain it. The above PR should fix the issue for you.

Just a warning though, decrypting and processing non-MDC messages (such as the ones you have) is not recommended and possibly insecure depending on where the messages come from. The https://www.efail.de/ vulnerabilities are closely related to this. If possible please upgrade whatever software you're using to generate the messages.

@spearmootz Do you have that config option set too? If not yours is probably a different issue.

**Comment by vballa** [comment]
> @vballa Ok, that does explain it. The above PR should fix the issue for you.
> 
> Just a warning though, decrypting and processing non-MDC messages (such as the ones you have) is not recommended and possibly insecure depending on where the messages come from. The https://www.efail.de/ vulnerabilities are closely related to this. If possible please upgrade whatever software you're using to generate the messages.
> 
> @spearmootz Do you have that config option set too? If not yours is probably a different issue.
> 
> If you do, it's probably the same issue but combined with something else that makes it inconsistent, perhaps some kind of file corruption making it seem like they're non-MDC messages when in fact they're not?

Thanks @twiss. Unfortunately the software that is generating message is not in our control. hence i had to go and set this option. 

May I know when are you planning to merge the PR? 

:+1: Thanks once again and hope this change should fix it.

**Comment by spearmootz** [comment]
i dont know if this is related but, i commented out that line and left it running in production, today i have another error. but when i restart the process it decrypts the file right away.

Error decryptingmessage: Unexpected end of packet
in line 35576:21

**Comment by twiss** [comment]
@vballa Next release should be early next week.

@spearmootz From `folderWatcher`, I gather that some other application is writing to a file, which you're then reading and decrypting, yes? Could it be that sometimes, when you read the file, the other application is not done writing yet? That would explain the "Unexpected end of packet". I'm not sure how it would cause "Cannot cancel a stream that already has a reader", though.

**Comment by spearmootz** [comment]
I'll check my code but I believe that I have built a module to wait for the
file to be finished writting. But I'll see

**Comment by korysas** [comment]
@twiss Hi I have come across this error when using streams during decryption. It only occurs when I set `openpgp.config.ignoreMdcError = true`.

Previous comments above indicate that this should be fixed even when using the ignore MDC option. Unfortunately, I need to set this option because I don't have control of the encryption process.

I'm using `v5.0.0-0` of the package. Please find my code sample below. Interestingly, the decryption works and in the end the decrypted file is written out. However, there is till the UnahndledPromiseRejectionWarning.

```
const decrypt = async function(encryptedFile) {
    const encryptedData = fs.createReadStream(process.env.ENCRYPTED_FILE_PATH);
    const privateKeyBlock = fs.readFileSync(process.env.PRIVATE_KEY_FILE_PATH);
    const privateKey = await openpgp.readKey({ armoredKey: privateKeyBlock });
    await privateKey.decrypt(process.env.PASS_PHRASE);

    const decrypted = await openpgp.decrypt({
        message: await openpgp.readMessage({ binaryMessage: encryptedData }),
        privateKeys: [privateKey],
        format: 'binary'
    });

    const readStream = decrypted.data;
    const writeStream = fs.createWriteStream(process.env.DECRYPTED_FILE_PATH, { flags: 'w' });
    readStream.pipe(writeStream);

    await new Promise((resolve) => {
        readStream.on('end', () => {
            console.log('done');

            resolve();
        });
    });
}
```

**Comment by korysas** [comment]
Thanks @twiss for the quick change! Do you know when the next release will be?

**Comment by twiss** [comment]
Hey @korysas, thanks for the report! I've released [v5.0.0-1](https://github.com/openpgpjs/openpgpjs/releases/tag/v5.0.0-1). We've also (unrelatedly) renamed `openpgp.config.ignoreMdcError` to `allowUnauthenticatedMessages`, and added some explanation of the risks - which you may already be aware of, but just be sure to warn your users of them as well, if any apply :blush: Don't process the decrypted data further, don't send it back to the sender, all good [efail](https://efail.de/) stuff ^.^
